DESIGN.md

```diff
@@ -0,0 +1,219 @@
+via the `sk.extensions.*` namespace.
+
+### `./memory` (Partial)
+
+This was a complex sub-system to port. The C# code has lots of interfaces and nesting
+of types and generics... in Python, I think we can simplify this a lot. An embedding
+is an `ndarray`. There's lots of great pre-built features that come with that. The
+rest of the system is a pretty direct port but I will say that the layering here
+is confusing. What's the real difference between storage, memory, memory record,
+data entry, an embedding, a collection, etc.? 
+
+#### TODOs
+
+Review of this subsystem. Lots of good testing. Maybe some kind of overview 
+documentation about the design. Maybe a diagram of how all these classes and interfaces
+fit together? I've got end-to-end examples with memory working but I still wouldn't 
+quite say I fully understand the arch. of the sub-system here.
+
+### `./orchestration` (Complete?)
+
+This was a pretty core piece and another direct port. I think we have pretty much
+everything in here? Worth double checking. Needs good docs and tests.
+
+### `./planning` (TODO: nothing yet)
+
+Completely ignored planning for now (and, selfishly, planning isn't a priority for 
+SK-based experimentation).
+
+### `./reliability` (Complete?)
+
+Direct port. Nothing much going on in this sub-module. Likely could use more strategies
+for retry. Also wasn't quite sure if this was integrated with the kernel/backends? 
+(Like are we actually using the re-try code, or is it not hit)
+
+#### TODOs
+
+Implement a real retry strategy that has backoff perhaps. Make sure this code is integrated
+and actually in use.
+
+### `./semantic_functions` (Complete?)
+
+Another core piece. The different config classes start to feel cumbersome here 
+(func config, prompt config, backend config, kernel config, so so much config).
+
+### `./semantic_functions/partitioning` (TODO: nothing yet)
+
+Skipped this sub-sub-module for now. Good task for someone to pick up!
+
+### `./skill_definition` (Complete?)
+
+Another core piece, another pretty direct port. 
+
+**NOTE:** the attributes in C# become decorators in Python. We probably could 
+make it feel a bit more pythonic (instead of having multiple decorators have just
+one or two). 
+
+**NOTE:** The skill collection, read only skill collection, etc. became a bit 
+confusing (in terms of the relationship between everything). Would be good to 
+double check my work there.
+
+### `./template_engine` (Complete?)
+
+Love the prompt templates! Have tried some basic prompts, prompts w/ vars,
+and prompts that call native functions. Seems to be working.
+
+**NOTE:** this module definitely needs some good tests. I could see some
+subtle errors sneaking into the prompt tokenization/rendering code here.
+
+### `./text` (TODO: nothing yet)
+
+Ignored this module for now.
+
+### `<root>` (Partial)
+
+I have a working `Kernel` and a working `KernelBuilder`. The base interface
```

Comment:
suggestion: throughout the doc I would avoid "I" which sounds like it will continue to be a personal project. I would suggest using generic statements about the software, what is ready and what not.